compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs

```diff
@@ -1357,6 +1357,18 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             ),
         };
         err.span_note(constraint_span, &msg);
+        if let ConstraintCategory::CallArgument = category {
+            if let Some(generator_kind) = use_span.generator_kind() {
+                if let GeneratorKind::Async(_) = generator_kind {
+                    err.note(
+                        "borrows cannot be held across a yield point, because the stack \
+                        space of the current function is not preserved",
```

Comment:
In retrospect I'm not sure this is actually correct, since it works as long as the executor isn't multi-threaded. I'm still a little hazy on async borrows work, maybe we should go with @tmandry's explanation instead:

> note: async blocks are not executed immediately and either must take a reference or ownership of outside variables they use

Reply:
Sorry, I actually don't understand that error message from @tmandry. How else could you use a variable, if not by reference or ownership? I'm contributing to the compiler in hopes of furthering my understanding of Rust. I find that the rules for the borrow checker are apparently simple, but because borrowing is so pervasively used throughout Rust, there are non obvious consequences, especially around things like closures (I actually haven't used async that much, but looking into the code, it seems like async and closures are somewhat similar regarding borrowing).

Reply:
@tmandry's message is trying to succinctly explain that `async` blocks need to be turned into a struct internally that owns or directly references every binding it touches coming from outside the block. The wording as it stands isn't very "newcomer friendly", but I find it hard to improve on its use of jargon, and for people that don't understand it the link to the async book should still be helpful.

Reply:
I see the confusion. Maybe just this, plus the help text, is enough to get the point across?

> note: async blocks are not executed immediately
> help: to force the async block to take ownership of `room_ref` (and any other referenced variables), use the `move` keyword

or something like

> note: async blocks are not executed immediately and by default refer to captured variables by reference

which seems a bit long.

Reply:
Thanks, @estebank, your explanation of the error from @tmandry clears it up for me. I will use the original explanation.